proposals/type-applications-in-patterns.rst

```diff
@@ -0,0 +1,170 @@
+Type Applications in Patterns
+=============================
+
+.. proposal-number::
+.. trac-ticket::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/126>`_.
+.. sectnum::
+.. contents::
+
+We have ``TypeApplications`` in terms. This proposal brings them to patterns in a way that preserves analogy to type signatures.
+
+
+Motivation
+------------
+
+``TypeApplications`` are a convenient and natural way to specifying types of polymorphic functions. Consider::
+
+ data Foo a where MkFoo :: forall a. a -> Foo a
+ 
+With ``TypeApplications``, I can replace the somewhat clumsy ``MkFoo (x :: ty)`` with ``MkFoo @ty x``. Seen this way,
+explicit type applications are merely alternative syntax for type signatures.
+
+At the moment, this only works in terms, but not in patterns: We can use type signatures in patterns
+(if ``PatternSignatures`` or ``ScopedTypeVariables`` are enabled), but not type applications. Given the strong
+relation between these syntactic forms, this is odd – why can I write::
+
+    foo (MkFoo (x :: ty)) = …
+   
+but not::
+
+    foo (MkFoo @ty x) = …
+
+This proposal fills this gap, by allowing type applications in pattern syntax and specifying it to behave
+“just like type signatures”.
+
+The intention of the following specification is that the following holds: For a constructor with type ``C :: forall a. a -> …`` the meaning of ``C @ty x`` should coincide with the existing form ``C (x :: ty)``. 
+
+Proposed Change Specification
+-----------------------------
+
+When both ``TypeApplications`` and ``PatternSignatures`` are enabled, then type application syntax is
+available in patterns. 
+
+If ``ScopedTypeVariables`` is enabled, then type applications and type signatures in pattern may mention type variables.
+
+A type variable mentioned in a pattern (in a type signature or a type application) that is not already in scope is brought into scope. The scope spans the pattern and the corresponding right-hand side.
+
+Typechecking a type application in a pattern follows the same rules as typechecking a pattern signature. This proposal does not propose any changes to the mechanism. In particular
```

Comment:
better "same rules as the existing rules for typechecking a pattern signature"

And what are those rules?  Although convenient, it's unsatfisfying to specify something by saying that it's similar to something else, also ill-specified.

Reply:
Well, this is a proposal that needs to specify a change to the language (“Proposed Change Specification”), so it seems prudent to point out deltas (or lack of deltas). A user manual would be written differently – but it would not discuss this feature in isolation, but rather give a coherent and consistent treatment of pattern matching in general.

Will add a link to the relevant section of the manual, though.